qiskit/transpiler/passes/routing/commuting_2q_gate_routing/commuting_2q_gate_grouper.py

```diff
@@ -0,0 +1,120 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2022.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""This pass collapses 2-qubit gates that commute into :class:`.Commuting2qBlock`."""
+
+from qiskit.circuit import Qubit, Gate
+from qiskit.dagcircuit import DAGCircuit, DAGOpNode
+from qiskit.transpiler.passes.optimization.commutation_analysis import CommutationAnalysis
+from qiskit.transpiler import TransformationPass, TranspilerError
+
+from qiskit.transpiler.passes.routing.commuting_2q_gate_routing.commuting_2q_block import (
+    Commuting2qBlock,
+)
+
+
+class Commuting2qGateGrouper(TransformationPass):
+    """A pass that collapses 2-qubit gates that commute into :class:`.Commuting2qBlock`."""
+
+    def __init__(self):
+        super().__init__()
+        self._commutation_set_by_node = None
+        self._done_nodes = set()
+        self.requires = [CommutationAnalysis()]
+
+    def run(self, dag: DAGCircuit) -> DAGCircuit:
+        """Collapses 2-qubit gates that commute into :class:`.Commuting2qBlock`.
+
+        Args:
+            dag: A DagCircuit in which commutation analysis was performed.
+
+        Returns:
+            A dag where multi-qubit commuting groups are collapsed as :class:`.Commuting2qBlock`.
+
+        Raises:
+            TranspilerError: If `commutation_set` is not in the property set.
+        """
+        if self.property_set["commutation_set"] is None:
+            raise TranspilerError("commutation_set is not in the property set.")
+
+        new_dag = dag.copy_empty_like()
+
+        for nd in dag.layered_topological_op_nodes():
```

Comment:
Instead of creating a new dag and iteratively adding all the operations to it like this, did you consider do an in place modification using `replace_block_with_op`? We do this in consolidate blocks (which seems very similar to this pass): https://github.com/Qiskit/qiskit-terra/blob/main/qiskit/transpiler/passes/optimization/consolidate_blocks.py#L116 I'm not sure which would have better performance but it's worth looking at.